iwyu_output.cc

```diff
@@ -1120,22 +1120,32 @@ void ProcessFullUse(OneUse* use,
     }
   }
 
-  // (B3) Discard symbol uses for builtin symbols, including new/delete.
-  // TODO(csilvers): we could use getBuiltinID(), but it returns
-  // non-zero for things like malloc.  Figure out how to use it.
-  if (const FunctionDecl* fn_decl = DynCastFrom(use->decl())) {
-    if (StartsWith(use->symbol_name(), "__builtin_")) {
-      VERRS(6) << "Ignoring use of " << use->symbol_name()
-               << " (" << use->PrintableUseLoc() << "): built-in function\n";
-      use->set_ignore_use();
-      return;
-    }
-    const string dfn_file = GetFilePath(fn_decl);
-    if (IsDefaultNewOrDelete(fn_decl, ConvertToQuotedInclude(dfn_file))) {
-      VERRS(6) << "Ignoring use of " << use->symbol_name()
-               << " (" << use->PrintableUseLoc() << "): built-in new/delete\n";
+  // (B3) Discard symbol uses for builtin symbols, including new/delete and
+  // template builtins.
+  if (isa<clang::BuiltinTemplateDecl>(use->decl())) {
+    VERRS(6) << "Ignoring use of " << use->symbol_name() << " ("
+             << use->PrintableUseLoc() << "): built-in template\n";
+    use->set_ignore_use();
+    return;
+  } else {
+    // A compiler builtin without a predefined header file (e.g. __builtin_..)
+    auto const iden = use->decl()->getIdentifier();
+    if (iden && iden->getBuiltinID() != 0 &&
+        !clang::Builtin::Context::isBuiltinFunc(use->symbol_name().c_str())) {
```

Comment:
This looks inverted -- shouldn't the `!` go away there? And does that mean there's a missing test?

Reply:
I think it is a misnomer since in both cases they will be functions, but if `isBuiltinFunc` returns true the function is one of the special __builtin_* that have no header or declaration anywhere. That's why I check that it has a valid builtin ID but it isn't one of this special ones. I assume that means it is then a regular libc/libm function (possibly with associated header) like malloc. That is also the reason why the comment said before that checking whether getBuiltinID returns > 0 is not enough.

For more info look in clang/Basic/Builtins.def where they are [defined](https://github.com/llvm-mirror/clang/blob/a3ba348aeea36f4cd6155854364e088bfd66d14d/include/clang/Basic/Builtins.def#L76). 'F' attribute means special builtin, 'f' means regular function. Or so I understand it :)

Ideally, I would have used `isPredefinedLibFunction` which does exactly what I want but it requires us to get hold of the *Context* object from the *CompilerInstance* somehow. I thought it wouldn't be worth it to go that route, but I am open for suggestions!

Reply:
Either I'm misunderstanding your explanation or you've confused something above. As far as I can tell `isBuiltinFunc` returns true for any `f`-marked builtin, i.e. the ones that look like normal functions and require a header (e.g. memcpy). So your code is doing the right thing (if identifier has a non-zero builtin and name does not appear in libbuiltin-table, it must be a proper `__builtin_` builtin).

I think if we wanted to be fancy, we could do this lookup when reporting the use, and encode it as a use-flag, but I think this is fine as-is.